### DISCUSSION TITLE
Custom dialect - Forcing quoting of all columns in an insert statement

### DISCUSSION THREAD

**Discussion by scottboston** [sqlalchemy/sqlalchemy#10647] [Usage Questions]
I have a custom dialect for a lakehouse database.

When doing Table.Insert the sql statement generated by sqlalchemy is not quoting all the columns,

I am using pandas.dataframe.to_sql with engine dialect to create table and insert data.

For example, the current generate statement is:

`INSERT INTO """ECS S3"".""power"".""lakehose_test""".table_one_test (date, col1, "col 2") VALUES (:date, :col1, :col_2)
`
Note "col 2" is quoted because of the space.  However, date needs to be quoted because this is a reserved word in this dialect, but properly, in my opinion, I would like to enable quoting of all columns. To have sqlalchemy generate 

`INSERT INTO """ECS S3"".""power"".""lakehose_test""".table_one_test ("date", "col1", "col 2") VALUES (:date, :col1, :col_2)
`
Is there a parameter or attribute I can set in the engine or dialect that will quote all columns?

**Comment by zzzeek** [comment]
Firstly, "date" should be in the reserved words that your dialect exports.  that will cause it to be quoted when used as an identifier name.  for example see https://github.com/gordthompson/sqlalchemy-access/blob/902dc166fce717ea6464ea3e89354b22f96da2e1/sqlalchemy_access/base.py#L362 in the sqlalchemy-access third party dialect which sets these up as part of its custom `IdentifierPreparer` class (which you will want to add to your custom dialect).

Secondly, the reason identifiers are not quoted universally is that this causes those identifiers to be used in a case-sensitive fashion, SQLAlchemy tries to avoid this for cases such as where the actual database tables already exist and might have different casings, but since this is a custom dialect you can do whatever you want.

Thirdly, why quote all column identifiers but not table identifiers?   you likely want to look at *identifier* casing overall.

to control identifier casing, within your `IdentifierPreparer` class override the [quote](https://github.com/sqlalchemy/sqlalchemy/blob/55bf2bf971cda82c2e3f58e9c211c74c3a6394a8/lib/sqlalchemy/sql/compiler.py#L7379) method. that's where the entire set of decisions takes place.
